### PR TITLE
Added Iterative Splitter to minimize memory usage when creating folds.

### PR DESCRIPTION
**Pull request by afcarvalho1991**
Implemented CrossValidationSplitterIterative<U, I>
Created a test to assess if the modifications are valid
CrossValidatedIterativeMahoutKNNRecommenderEvaluator
In CrossValidationSplitter, the visibility of the following variables
was change to protected:
1 - nFolds;
2 - perUser;
3 - rdn.
Modified TemporalDataModel constructor userItemTimestamps was not
initialized. For this reason, there are commented sections in CrossValidationSplitterIterative<U, I> to allow the execution of the testing.
Results from CrossValidatedIterativeMahoutKNNRecommenderEvaluator:
NDCG@10: 0.02932215503411596
RMSE: 1.1055677436810647
P@10: 0.04110286320254521

BUG from previous version (master) -> Need to be fixed->
CrossValidatedMahoutKNNRecommenderEvaluator not working, see Log below.

----------------------------- LOG --------------------------------------

5 [main] INFO org.apache.mahout.cf.taste.impl.model.file.FileDataModel -
Creating FileDataModel for file data/ml-100k/model/train_0.csv
52 [main] INFO org.apache.mahout.cf.taste.impl.model.file.FileDataModel

Reading file info...
Exception in thread "main" java.lang.NumberFormatException: For input
string: ""
at
java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
at java.lang.Long.parseLong(Long.java:601)
at java.lang.Long.parseLong(Long.java:631)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.readTimestampFromString(FileDataModel.java:632)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.addTimestamp(FileDataModel.java:595)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.processLine(FileDataModel.java:512)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.processFile(FileDataModel.java:340)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.buildModel(FileDataModel.java:236)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.reload(FileDataModel.java:208)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.(FileDataModel.java:194)
at
org.apache.mahout.cf.taste.impl.model.file.FileDataModel.(FileDataModel.java:149)
at
net.recommenders.rival.examples.movielens100k.CrossValidatedMahoutKNNRecommenderEvaluator.recommend(CrossValidatedMahoutKNNRecommenderEvaluator.java:160)
at
net.recommenders.rival.examples.movielens100k.CrossValidatedMahoutKNNRecommenderEvaluator.main(CrossValidatedMahoutKNNRecommenderEvaluator.java:93)
------------------------------- end-LOG --------------------------------

André Carvalho

### CODE REVIEW

File: rival-examples/src/main/java/net/recommenders/rival/examples/movielens100k/CrossValidatedIterativeMahoutKNNRecommenderEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+//            System.out.println("train: " + trainingFile);
+//            System.out.println("test: " + testFile);
+//            boolean overwrite = true;
+//            try {
+//                DataModelUtils.saveDataModel(training, trainingFile, overwrite, "\t");
+//                DataModelUtils.saveDataModel(test, testFile, overwrite, "\t");
+//            } catch (FileNotFoundException | UnsupportedEncodingException e) {
+//                e.printStackTrace();
+//            }
+//        }
```
Comment (alansaid) [inline rival-examples/src/main/java/net/recommenders/rival/examples/movielens100k/CrossValidatedIterativeMahoutKNNRecommenderEvaluator.java]: Please remove old code chunks if they are not necessary anymore

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+//import java.util.Set;
+
+//import net.recommenders.rival.core.DataModel;
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: clean up old imports if no longer needed

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	            int n = 0;
+	            for (U user : data.getUsers()) {
+	                List<I> items = new ArrayList<I>(data.getUserItemPreferences().get(user).keySet());
+	                Collections.shuffle(items, rnd);
+	                for (I item : items) {
+	                    Double pref = data.getUserItemPreferences().get(user).get(item);
+//	                    Set<Long> time = null;
+//	                    if (data.getgetUserItemTimestamps().containsKey(user) && data.getUserItemTimestamps().get(user).containsKey(item)) {
+//	                        time = data.getUserItemTimestamps().get(user).get(item);
+//	                    }
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: Remove unnecessary comments

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	                        if(f_writer==null) continue; // not a "valid" fold already computed
+	                        
+	                        if (pref != null) {
+	                        	f_writer.write(user+"\t"+item+"\t"+pref);
+	                        }
+//	                        if (time != null) {
+//	                            for (Long t : time) {
+//	                            	f_writer.write("\t"+t);
+//	                            }
+//	                        }
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: if not necessary, then remove

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	            for (U user : users) 
+	            {
+	                List<I> items = new ArrayList<I>(data.getUserItemPreferences().get(user).keySet());
+	                Collections.shuffle(items, rnd);
+	                for (I item : items) {
+	                    Double pref = data.getUserItemPreferences().get(user).get(item);
+//	                    Set<Long> time = null;
+//	                    if (data.getUserItemTimestamps().containsKey(user) && data.getUserItemTimestamps().get(user).containsKey(item)) {
+//	                        time = data.getUserItemTimestamps().get(user).get(item);
+//	                    }
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: if not necessary, remove

File: rival-examples/src/main/java/net/recommenders/rival/examples/movielens100k/CrossValidatedIterativeMahoutKNNRecommenderEvaluator.java [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import net.recommenders.rival.core.DataModelFactory;
+import net.recommenders.rival.evaluation.metric.error.RMSE;
+
+/**
+ * RiVal Movielens100k Mahout Example, using 5-fold cross validation.
+ *
+ * @author <a href="http://github.com/alansaid">Alan</a>
+ */
```
Comment (alansaid) [inline rival-examples/src/main/java/net/recommenders/rival/examples/movielens100k/CrossValidatedIterativeMahoutKNNRecommenderEvaluator.java]: Update JavaDoc to reflect what's going on

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+//import java.util.Set;
+
+//import net.recommenders.rival.core.DataModel;
+import net.recommenders.rival.core.DataModelIF;
+
+public class CrossValidationSplitterIterative<U, I>  extends CrossValidationSplitter<U, I>
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: Stick to the naming convention, i.e. IterativeCrossValidationSplitter

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
+package net.recommenders.rival.split.splitter;
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: Insert license notice at top

File: rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+//import java.util.Set;
+
+//import net.recommenders.rival.core.DataModel;
+import net.recommenders.rival.core.DataModelIF;
+
```
Comment (alansaid) [inline rival-split/src/main/java/net/recommenders/rival/split/splitter/CrossValidationSplitterIterative.java]: Add JavaDoc for class

**Comment by abellogin** [comment]
Hi,

regarding the error in CrossValidatedMahoutKNNRecommenderEvaluator, it seems it was generated in commit e9f996f, where a "-1" was deleted after the last delimiter in DataModelUtils.

I will create an issue now.

Thanks,
Alex

**Comment by afcarvalho1991** [comment]
@alansaid I will revise the code when the CrossValidatedMahoutKNNRecommenderEvaluator is fixed, since those commented sections avoid the runtime error when executing CrossValidatedIterativeMahoutKNNRecommenderEvaluator. 

Tomorow I will have a look and provide an updated version with your requests. 

André.

**Comment by abellogin** [comment]
Hi André,

as you may already know, that class is fixed (see #124 ), so, please, update your code accordingly and the pull request (I am not sure how the reviews work, tho) whenever you have the time.

**Comment by alansaid** [comment]
Thanks @afcarvalho1991!

**Comment by afcarvalho1991** [comment]
Glad that I could help!

**Pull request merged**